src/target/stm32mp15.c

```diff
@@ -0,0 +1,154 @@
+/*
+ * This file is part of the Black Magic Debug project.
+ *
+ * Copyright (C) 2023 1BitSquared <info@1bitsquared.com>
+ * Written by ALTracer <tolstov_den@mail.ru>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * This file implements STM32MP15 target specific functions for detecting
+ * the device and providing the XML memory map.
+ *
+ * References:
+ * RM0436 - STM32MP157 advanced Arm®-based 32-bit MPUs, Rev. 5
+ *   https://www.st.com/resource/en/reference_manual/rm0436-stm32mp157-advanced-armbased-32bit-mpus-stmicroelectronics.pdf
+ */
+
+#include "general.h"
+#include "target.h"
+#include "target_internal.h"
+#include "cortexm.h"
+
+/* Access from processor address space.
+ * Access via the debug APB is at 0xe0081000 over AP1. */
+#define DBGMCU_IDCODE      0x50081000U
+#define STM32MP15_UID_BASE 0x5c005234U
+
+#define DBGMCU_IDC (DBGMCU_IDCODE + 0U)
+#define DBGMCU_CR  (DBGMCU_IDCODE + 4U)
+#define DBGSLEEP   (1U << 0U)
+#define DBGSTOP    (1U << 1U)
+#define DBGSTBY    (1U << 2U)
+
+#define ID_STM32MP15x        0x5000U /* RM0436 */
+#define ID_STM32MP15x_ERRATA 0x450U
+
+typedef struct stm32mp15_priv {
+	uint32_t dbg_cr;
+} stm32mp15_priv_s;
+
+static bool stm32mp15_uid(target_s *target, int argc, const char **argv);
+static bool stm32mp15_cmd_rev(target_s *target, int argc, const char **argv);
+
+const command_s stm32mp15_cmd_list[] = {
+	{"uid", stm32mp15_uid, "Print unique device ID"},
+	{"revision", stm32mp15_cmd_rev, "Returns the Device ID and Revision"},
+	{NULL, NULL, NULL},
+};
+
+static void stm32mp15_detach(target_s *target)
+{
+	stm32mp15_priv_s *ps = (stm32mp15_priv_s *)target->target_storage;
+	target_mem_write32(target, DBGMCU_CR, ps->dbg_cr);
+	cortexm_detach(target);
+}
+
+bool stm32mp15_probe(target_s *target)
```

Comment:
Given this is for the Cortex-M core, and there are Cortex-A cores in this part that will need their own, separate, probe routine, we suggest including that information in the probe routine's name - `stm32mp15_cm4_probe`. That way there's room left for a `stm32mp15_ca7_probe` or similar without having to come back and rename this.

Reply:
Renamed as suggested. Though the CA7 is only accessible in Engineering Boot for boot chain debugging (BootROM traces, TF-A BL2 or U-Boot SPL, TF-A SP-MIN or OP-TEE) during platform bring-up, and Linux kernel is often debugged by its KDB/KGDB stub over UART.
This item was the first thing I did revisiting this PR, but apparently I lost the first commit in this chain when moving patches around. Re-done.